xdsl/interpreter.py

```diff
@@ -0,0 +1,273 @@
+                # This is an annotated function
+                op_type = getattr(val, _IMPL_OP_TYPE)
+                if op_type not in impl_dict:
+                    # subclass overrides superclass definition
+                    impl_dict[op_type] = val  # type: ignore
+    setattr(ft, _IMPL_DICT, impl_dict)
+
+    @classmethod
+    def impls(
+        cls: type[_FT]
+    ) -> Iterable[tuple[type[Operation], OpImpl[InterpreterFunctions,
+                                                Operation]]]:
+        return impl_dict.items()
+
+    setattr(ft, 'impls', impls)
+    return ft
+
+
+@dataclass
+class _InterpreterFunctionImpls:
+    """
+    Used to combine multiple function implementations. The operation
+    implementations need to be passed the instance of the Functions class,
+    so we keep a `(Functions, OpImpl)` tuple for every Operation type.
+    """
+
+    _impl_dict: dict[type[Operation],
+                     tuple[InterpreterFunctions,
+                           OpImpl[InterpreterFunctions,
+                                  Operation]]] = field(default_factory=dict)
+
+    def register_from(self, ft: InterpreterFunctions, /, override: bool):
+        for op_type, impl in ft.impls():
+            if op_type in self._impl_dict and not override:
+                raise ValueError(
+                    "Attempting to register implementation for op of type "
+                    f"{op_type}, but type already registered")
+
+            self._impl_dict[op_type] = (ft, impl)
+
+    def run(self, interpreter: Interpreter, op: Operation,
+            args: tuple[Any, ...]) -> tuple[Any, ...]:
+        if type(op) not in self._impl_dict:
+            raise InterpretationError(
+                f'Could not find interpretation function for op {op.name}')
+        ft, impl = self._impl_dict[type(op)]
+        return impl(ft, interpreter, op, args)
+
+
+@dataclass
+class InterpreterContext:
+    """
+    Class holding the Python values associated with SSAValues during an interpretation
+    context. An environment is a stack of scopes, values are assigned to the current
+    scope, but can be fetched from a parent scope.
+    """
+
+    name: str = field(default="unknown")
+    parent: InterpreterContext | None = None
+    env: dict[SSAValue, Any] = field(default_factory=dict)
+
+    def __getitem__(self, key: SSAValue) -> Any:
+        """
+        Fetch key from environment. Attempts to first fetch from current scope, then
+        from parent scopes. Raises Interpretation error if not found.
+        """
+        if key in self.env:
+            return self.env[key]
+        if self.parent is not None:
+            return self.parent[key]
+        raise InterpretationError(f'Could not find value for {key} in {self}')
+
+    def __setitem__(self, key: SSAValue, value: Any):
+        """
+        Assign key to current scope. Raises InterpretationError if key already assigned to.
+        """
+        if key in self.env:
```

Comment:
My feeling is that stacks are only used for function calls, right?
So I think that we should allow to overwrite values, for instance in arbitrary CFG later, or for `scf.for` for instance?

Reply:
That's not quite the case, I use the stack as a kind of hack in the PDL interpreter, and I can imagine it being used in some other interesting ways in the future. I'd leave it flexible for now.

Reply:
Yes we should keep the `stack`, but we should remove this error, right?
Otherwise, how could we implement something like a `for` loop? Which will overwrite values.